Support for killing individual task if cluster is out of memory

losipiuk opened the pull request:
<!-- Thank you for submitting a pull request! Find more information in our development guide at https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md and contact us on #dev in Slack. -->

## Description

Support for killing individual task if cluster is out of memory.
Individual tasks will be killed for queries which are run with task-level retries.
Currently only supported if default low memory killer policy is used (total-reservation-on-blocked-nodes).

> Is this change a fix, improvement, new feature, refactoring, or other?

improvement to new feature

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

core query engine

> How would you describe this change to a non-technical end user or system administrator?

N/A

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Section
* Add support for killing individual tasks of queries run with ``retry-mode`` set to ``TASK`` low memory killer.
  Requires ``query.low-memory-killer.policy`` config option to be set to ``total-reservation-on-blocked-nodes``. ({issue}`11129`)
```

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/RemoteTask.java:
```diff
@@ -65,6 +65,8 @@
 
     void fail(Throwable cause);
 
+    void failWaitForUpdate(Throwable cause);
```
nit: maybe think about a better name (though It's difficult for me to find a better one)

losipiuk replied:
Will rename to failRemotely

arhimondr commented on core/trino-main/src/main/java/io/trino/execution/scheduler/SqlQueryScheduler.java:
```diff
@@ -475,6 +475,17 @@ public synchronized void abort()
+    public void failTask(TaskId taskId, Throwable failureCause)
+    {
+        try (SetThreadName ignored = new SetThreadName("Query-%s", queryStateMachine.getQueryId())) {
+            coordinatorStagesScheduler.failTask(taskId, failureCause);
```
nit: maybe it is better to be consistent

arhimondr commented on core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java:
```diff
@@ -317,7 +415,57 @@ private boolean isClusterOutOfMemory()
+        return getRetryPolicy(query.getSession()) == RetryPolicy.TASK;
+    }
+
+    private Multimap<String, TaskMemoryInfo> createTaskMemoryInfos(QueryExecution query)
```
might be an artifact from the old implementation (since you already send the memory reservation as part of the memory reservation response)

arhimondr reviewed (commented):
Looks good to me % comments

losipiuk commented:
AC

losipiuk commented:
@martint could you maybe take a look?

losipiuk commented:
Major changes:
* Rebased on top of https://github.com/trinodb/trino/pull/11129.
* Dropped mechanism of per-task memory limit. We are not enforcing limits right now. We are estimating tasks size, and bin-packing based on that, but we allow it to go above. If memory pool is exhaust memory kill will intervene and kill one of the tasks
* I dropped integration test. With memory-killer approach I was not able to make it work (deterministically enforce that one of the tasks would require full not allocation)

@arhimondr PTAL once again

linzebing commented on core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java:
```diff
@@ -228,11 +230,11 @@ public synchronized void process(Iterable<QueryExecution> runningQueries, Suppli
             }
             else {
-                log.debug("Last killed query is still not gone: %s", lastKilledQuery);
+                log.debug("Last killed query is still not gone: %s", lastKillTarget);
```
```suggestion
                log.debug("Last killed target is still not gone: %s", lastKillTarget);
```

linzebing commented on core/trino-main/src/main/java/io/trino/memory/ClusterMemoryManager.java:
```diff
@@ -245,38 +247,72 @@ private synchronized void callOomKiller(Iterable<QueryExecution> runningQueries)
+                log.debug("Low memory killer chose %s", queryId);
+                Optional<QueryExecution> chosenQuery = findRunningQuery(runningQueries, killTarget.get().getQuery());
+                if (chosenQuery.isPresent()) {
+                    // See comments in  isLastKilledQueryGone for why chosenQuery might be absent.
```
```suggestion
                    // See comments in isQueryGone for why chosenQuery might be absent.
```

linzebing commented on core/trino-main/src/test/java/io/trino/memory/TestTotalReservationOnBlockedNodesLowMemoryKiller.java:
```diff
@@ -75,4 +78,58 @@ public void testSkewedQuery()
+        Map<String, Map<String, Map<String, Long>>> tasks = ImmutableMap.<String, Map<String, Map<String, Long>>>builder()
+                .put("q_2", ImmutableMap.of(
+                        "n1", ImmutableMap.of("t1", 1L, "t2", 3L),
+                        "n2", ImmutableMap.of("t3", 3L, "t4", 1L, "t5", 1L),
```
The total task reservation of q_2 on n2 doesn't equal to the query reservation of q_2 on n2, it it expected? I understand that it doesn't affect the effectiveness of the test, though.

losipiuk replied:
Actually, it matches: 3+1+1 = 5. But yeah - the killing logic does not care about total query reservation for queries with task level retries.
